flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperator.java

```diff
@@ -273,32 +284,93 @@ public void onProcessingTime(InternalTimer<K, W> timer) throws Exception {
 			return;
 		}
 
-		TriggerResult triggerResult = context.onProcessingTime(timer.getTimestamp());
-		if (triggerResult.isFire()) {
-			fire(context.window, contents);
-		}
+				TriggerResult triggerResult = context.onProcessingTime(timer.getTimestamp());
+				if (triggerResult.isFire()) {
+					fire(context.window, contents, windowState);
+				}
 
 		if (triggerResult.isPurge() || (!windowAssigner.isEventTime() && isCleanupTime(context.window, timer.getTimestamp()))) {
 			cleanup(context.window, windowState, mergingWindows);
 		}
 	}
 
-	private void fire(W window, Iterable<StreamRecord<IN>> contents) throws Exception {
+	private void fire(W window, Iterable<StreamRecord<IN>> contents, ListState<StreamRecord<IN>> windowState) throws Exception {
 		timestampedCollector.setAbsoluteTimestamp(window.maxTimestamp());
 
 		// Work around type system restrictions...
-		int toEvict = evictor.evict((Iterable) contents, Iterables.size(contents), context.window);
-
-		FluentIterable<IN> projectedContents = FluentIterable
+		FluentIterable<TimestampedValue<IN>> recordsWithTimestamp = FluentIterable
 			.from(contents)
-			.skip(toEvict)
-			.transform(new Function<StreamRecord<IN>, IN>() {
+			.transform(new Function<StreamRecord<IN>, TimestampedValue<IN>>() {
+				@Override
+				public TimestampedValue<IN> apply(StreamRecord<IN> input) {
+					return new TimestampedValue<>(input.getValue(), input.getTimestamp());
+				}
+			});
+		evictorContext.evictBefore(recordsWithTimestamp, Iterables.size(recordsWithTimestamp));
+
+		FluentIterable<IN> projectedContents = recordsWithTimestamp
+			.transform(new Function<TimestampedValue<IN>, IN>() {
 				@Override
-				public IN apply(StreamRecord<IN> input) {
+				public IN apply(TimestampedValue<IN> input) {
 					return input.getValue();
 				}
 			});
+
 		userFunction.apply(context.key, context.window, projectedContents, timestampedCollector);
+		evictorContext.evictAfter(recordsWithTimestamp, Iterables.size(recordsWithTimestamp));
+
+
+		//work around to fix FLINK-4369, remove the evicted elements from the windowState.
+		//this is inefficient, but there is no other way to remove elements from ListState, which is an AppendingState.
+		windowState.clear();
+		for(TimestampedValue<IN> record : recordsWithTimestamp) {
+			if (record.getTimestamp() < 0) {
```

Comment:
I think this does not work properly in all cases because negative timestamps are in fact also valid timestamps. You can look at `StreamRecord` for how this should be solved in `TimestampedValue` and also for how the copying between the two should work.